### PR TITLE
Implement Laplace (quadratic) approximation

### PR DESCRIPTION
**Pull request by carsten-j**
This is an early version of q quadratic approximation implementation that I have developed while reading Statistical Rethinking by Richard McElreath.

There is a short discussion about this in the [issue](https://github.com/pymc-devs/pymc-experimental/issues/341) and maybe @theorashid can help with feedback of this draft PR. 

This work is partly based on the Python package [pymc3-quap](https://github.com/rasmusbergpalm/pymc3-quap) but pymc3-quap is based on PYMC3 and a lot happend bewteen version 3 and 5 of PYMC. Optimizers works better when provided with a good initial guess and hence a (optional) starting point has been added to function arguments. Please see [Github](https://github.com/pymc-devs/pymc/issues/5443#issuecomment-1030609090) for a discussion about the differences between PYMC version 3 and 5 for computing the Hessian.

### CODE REVIEW

**Comment by carsten-j** [comment]
I am looking for the best way to return not just a posterior sample distribution but also the mean vector and covariance matrix of the Gaussian distribution. Any suggestion for this. So far my only idea is to add another section to the inferenceData returned containing this information. Thoughts on this?

**Comment by zaxtax** [comment]
I'm not sure the InferenceData is the best place to put it. We should copy
whatever we do with Variational Inference

**Comment by aloctavodia** [comment]
Historically, Inferencedata has been focused on mcmc. But we have discussed a few times extend it to better handle other inference methods, like SMC or variational methods. It just that there has not been enough momentum to agree and implement and schema that works for those methods.

**Comment by carsten-j** [comment]
@zaxtax and @aloctavodia are you saying that I should not return inferencedata at all or just not return the gaussian mean and covariance in the inferencedata object? I am new to both PYMC and Bayesian statistics so I do not know the history of this package.
Best, Carsten

**Comment by zaxtax** [comment]
Oh, it's more that we haven't decided how to handle this within the
library. Don't treat this as a blocker, though we should raise it for
discussion more broadly

**Comment by twiecki** [comment]
CC @ferrine

**Comment by ricardoV94** [comment]
Suggestion, have two groups in the returned inferencedata: 
1. A `fit` group that includes the mean and covariance of the laplace fit, and 
2. a `posterior` group that includes draws from this fit, and has all the bells and whistles like dimensions, deterministics, etc... Include the default extra groups like observed and constant data. This will look just like a fit from mcmc sampling. This can be disabled by the user by setting `draws = 0`

We could even try different fits from distinct initialization points (optionally) and save those as distinct "chains" in the `fit` and corresponding `posterior` groups. Although usually multiple initialization are used with the goal of finding the best fit, they could still be useful to detect multi-modality / pervasiveness of local optima.

File: pymc_experimental/inference/quadratic.py
Comment (ricardoV94) [inline pymc_experimental/inference/quadratic.py:87]: Use the utility `remove_value_transforms`: https://www.pymc.io/projects/docs/en/stable/api/model/generated/pymc.model.transform.conditioning.remove_value_transforms.html

This must be done outside of a model context for now, we'll fix that soon. That means for testing purposes you must call `pm.fit(..., model=model)` outside of a model context

File: pymc_experimental/inference/quadratic.py
Comment (ricardoV94) [inline pymc_experimental/inference/quadratic.py:77]: Nitpick: More standard abbrevation, corresponds to a more precise name
```suggestion
    rng = np.random.default_rng(seed=random_seed)
```

File: pymc_experimental/inference/quadratic.py
Comment (ricardoV94) [inline pymc_experimental/inference/quadratic.py:25]: Mention laplace somewhere early in the docstrings. Some folks have expressed a preference for calling it laplace actually (in which case quadratic should be mentioned early in the docstrings). I don't have a preference myself.

Comment (theorashid) [reply pymc_experimental/inference/quadratic.py:25]: I would prefer Laplace. numpyro also use Laplace https://docs.pyro.ai/en/1.4.0/infer.autoguide.html#autolaplaceapproximation

File: pymc_experimental/inference/quadratic.py
Comment (ricardoV94) [inline pymc_experimental/inference/quadratic.py:114]:
```suggestion
    trace.posterior = pm.compute_deterministics(trace.posterior, model=model, merge_dataset=True, progressbar=False)
```

File: pymc_experimental/inference/quadratic.py
Comment (ricardoV94) [inline pymc_experimental/inference/quadratic.py:79]: Allow controlling progressbar
```suggestion
    map = pm.find_MAP(vars=vars, progressbar=progressbar)
```

**Comment by ricardoV94** [comment]
@carsten-j PR looks great! I left some comment above

File: pymc_experimental/tests/test_quadratic.py
Comment (twiecki) [inline pymc_experimental/tests/test_quadratic.py:1]:
```suggestion
#   Copyright 2024 The PyMC Developers
```

**Comment by carsten-j** [comment]
Thanks you @ricardoV94 and @twiecki for the review comments. I believe that all of them expect one has been fixed. I have not figured out how to use `remove_value_transforms`. I tried to browse through PYMC source code but that did not really help.

**Comment by ricardoV94** [comment]
> Thanks you @ricardoV94 and @twiecki for the review comments. I believe that all of them expect one has been fixed. I have not figured out how to use `remove_value_transforms`. I tried to browse through PYMC source code but that did not really help.

The docs contains code example: https://www.pymc.io/projects/docs/en/stable/api/model/generated/pymc.model.transform.conditioning.remove_value_transforms.html

**Comment by carsten-j** [comment]
I should have mentioned that I did read the doc and looked at the example. But I have not been able to figure out how to apply it to my case. I will try again ...

**Comment by ricardoV94** [comment]
To be able to use it inside the model context, it will need this change to get merged first: https://github.com/pymc-devs/pymc/pull/7352

But you should be able to already test by doing the object way with `pm.fit(..., model=model)` outside of the model context

**Comment by carsten-j** [comment]
@ricardoV94 I figured out how to replace the for loop with remove_value_transforms. Is the PR ready for merge or are there additional review comments?

File: pymc_experimental/inference/laplace.py
Comment (ricardoV94) [inline pymc_experimental/inference/laplace.py:104]:
```suggestion
    transformed_m = pm.modelcontext(model)
    map = pm.find_MAP(vars=vars, progressbar=progressbar, model=transformed_m)
```

File: pymc_experimental/inference/laplace.py
Comment (ricardoV94) [inline pymc_experimental/inference/laplace.py:109]:
```suggestion
    untransformed_m = remove_value_transforms(transformed_m):
    untransformed_vars = [untransformed_m[v.name] for v in vars]
    hessian = pm.find_hessian(point=map, vars=untransformed_vars, model=untransformed_m)
```

**Review by ricardoV94 — CHANGES_REQUESTED** [review]
Does this work? If we already have the model, there's no need to enter the model context and force the inner functions to retrieve it again, we can just pass the model directly

**Comment by carsten-j** [comment]
@ricardoV94 it does work but I have updated the code according to your comment. It makes sense and I am still learning about PYMC :-)

File: pymc_experimental/inference/laplace.py
Comment (ricardoV94) [inline pymc_experimental/inference/laplace.py:42]:
```suggestion
    Create a Laplace (quadratic) approximation for a posterior distribution.
```

**Review by ricardoV94 — APPROVED** [review]
Looks pretty slick, left a minor docs suggestion

**Comment by carsten-j** [comment]
@ricardoV94 tests are failing but I do not see any relationship between the failing tests and this new Laplace feature. It there a general problem with tests?

**Comment by carsten-j** [comment]
@ricardoV94 tests are still failing. I am not sure what to do here as I do not see why these tests should fail due to my PR. Please advice.

**Comment by ricardoV94** [comment]
@carsten-j I can confirm those failures are unrelated to your changes, and related to changes in PyMC or other dependencies. We'll have to fix them in a separate PR

**Comment by carsten-j** [comment]
Thanks for clarifying @ricardoV94. What is next step from here. Do we need approval for additional maintainers or can the PR be merged now?

**Comment by ricardoV94** [comment]
@carsten-j we would fix the tests elsewhere, then update this PR on top of the fixes, and if the tests are passing we can merge, unless any of the reviewers blocks it / requests more changes. You don't need to do anything :)

**Comment by zaxtax** [comment]
Looks good. Once the tests pass, I think it's good to merge

File: pymc_experimental/inference/laplace.py
Comment (theorashid) [inline pymc_experimental/inference/laplace.py:57]: +returned.

File: pymc_experimental/tests/test_laplace.py
Comment (theorashid) [inline pymc_experimental/tests/test_laplace.py:44]: Question: say you only did `vars=[mu]`, how would the variable `logsigma` be estimated?

Comment (ricardoV94) [reply pymc_experimental/tests/test_laplace.py:44]: I think find_MAP in that case uses the initial_point for the excluded variable. I never found that behavior useful tbh

Comment (ricardoV94) [reply pymc_experimental/tests/test_laplace.py:44]: Sounds like we may want to pass a similar `start` kwarg to laplace to set the value of variables that are not being optimized?

File: pymc_experimental/inference/laplace.py
Comment (ricardoV94) [inline pymc_experimental/inference/laplace.py:56]: Only returning the `fit` should be draws=0?

Re: chains, right now it doesn't make much sense because all chains will be the same. This changes if: 1) Users can pass different starts for different chains or 2) We add some random jitter across the initial point so that multiple chains could indeed arrive at different conclusions. I'm not sure this is useful, so maybe limiting to 1 chain / optimization (and removing the kwarg) may be fine?

Comment (ricardoV94) [reply pymc_experimental/inference/laplace.py:56]: This was suggested by @theorashid

File: pymc_experimental/tests/test_laplace.py
Comment (theorashid) [reply pymc_experimental/tests/test_laplace.py:44]: worth adding a test on this to confirm the behaviour

File: pymc_experimental/inference/laplace.py
Comment (theorashid) [reply pymc_experimental/inference/laplace.py:56]: Just adding that it is pedagogically confusing to have multiple chains, because the user might try to incorrectly use MCMC diagnostics on the posterior to see if their "Laplace approximation" has converged. This is (probably) the same reason that the variation inference objects return idata with only one chain (the user again cannot specify multiple chains).

Comment (carsten-j) [reply pymc_experimental/inference/laplace.py:56]: I think it makes sense to limit to 1 chain and not having chains as part of the method arguments.

Comment (carsten-j) [reply pymc_experimental/inference/laplace.py]: The latest commit fixes this issue.

File: pymc_experimental/tests/test_laplace.py
Comment (carsten-j) [reply pymc_experimental/tests/test_laplace.py:43]: I am not sure I fully understand this. I will give it a second go with the documentation for find_MAP.

File: pymc_experimental/inference/laplace.py
Comment (theorashid) [inline pymc_experimental/inference/laplace.py:35]: nit: I think it would be `draws: Optional[int] = 1000` and then later down `if draws is not None:`

Comment (carsten-j) [reply pymc_experimental/inference/laplace.py]: This is fixed with latest commit

File: pymc_experimental/inference/laplace.py
Comment (carsten-j) [reply pymc_experimental/inference/laplace.py]: This has been fixed.

File: pymc_experimental/tests/test_laplace.py
Comment (theorashid) [reply pymc_experimental/tests/test_laplace.py:43]: Yeah, just a little confirmation of what happens in the test case when you don't pass all the variables into `pm.fit` e.g. `vars = [logsigma]`. Doesn't need to be a full test if it doesn't break things, it can be a warning or a line in the docs at this point.

Rest of the PR looks great. Nice job

Comment (carsten-j) [reply pymc_experimental/tests/test_laplace.py:43]: @theorashid it sound like a good idea with a warning in case one does not pass all variable into the fit method. I was going through the documentation and wondering if the PYMC model attribute 'value_vars' is the proper attribute to use to check if number of variables in 'vars' is the same as all the variables in the model.

Comment (carsten-j) [reply pymc_experimental/tests/test_laplace.py:43]: Hi @theorashid. I am not sure how to handle if only a subset of the variables are passed in, e.g. vars=[mu] and log_sigma is left out. If this should raise a warning I need some way of figuring out the number of model parameters and compare that with the number of parameters in vars. I am not sure how to determine the number of model parameters

Comment (ricardoV94) [reply pymc_experimental/tests/test_laplace.py:43]: `model.free_RVs`

Comment (carsten-j) [reply pymc_experimental/tests/test_laplace.py:43]: @theorashid and @ricardoV94, I committed an update that will raise a warning in case number of variables in vars does not equal number of model variables.

**Comment by ricardoV94** [comment]
@carsten-j tests are no longer failing in main. You can rebase/merge into your branch

File: pymc_experimental/tests/test_laplace.py
Comment (theorashid) [reply pymc_experimental/tests/test_laplace.py:43]: LGTM

**Comment by carsten-j** [comment]
@ricardoV94  I have rebased the laplace branch but it looks like someone needs to approve Github worksflows.

**Comment by zaxtax** [comment]
Looks like there are still a few failing tests, but once those pass this is probably good to merge

**Comment by carsten-j** [comment]
@zaxtax, all tests passed. Are you also able to merge the PR? Thanks.

**Comment by twiecki** [comment]
Congrats @carsten-j, this is a big one!

**Comment by carsten-j** [comment]
Thank you @twiecki. Really happy to contribute and thanks to all those that helped. After the summer I will try to work on documentation for building and running locally. I took me some time to figure out how this works!

**Comment by zaxtax** [comment]
Congrats @carsten-j this is really neat!

**Comment by theorashid** [comment]
Brilliant work @carsten-j . Hope to see you contribute to PyMC again!

**Pull request merged**
